File: frontend/demo/component/grid/grid-dynamic-height.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/button';
import '@vaadin/combo-box';
import type { ComboBoxValueChangedEvent } from '@vaadin/combo-box';
import '@vaadin/grid';
import { columnBodyRenderer } from '@vaadin/grid/lit.js';
import type { GridColumnBodyLitRenderer } from '@vaadin/grid/lit.js';
import '@vaadin/horizontal-layout';
import '@vaadin/icon';
import '@vaadin/icons';
import { getPeople } from 'Frontend/demo/domain/DataService';
import type Person from 'Frontend/generated/com/vaadin/demo/domain/Person';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('grid-dynamic-height')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  @state()
  private items: Person[] = [];

  @state()
  private invitedPeople: Person[] = [];

  @state()
  private selectedValue = '';

  protected override async firstUpdated() {
    const { people } = await getPeople();
    this.items = people.map((person) => ({
      ...person,
      displayName: `${person.firstName} ${person.lastName}`,
    }));
  }

  protected override render() {
    return html`
      <vaadin-horizontal-layout theme="spacing">
        <vaadin-combo-box
          .items="${this.items}"
          .value="${this.selectedValue}"
          item-label-path="displayName"
          item-value-path="id"
          style="flex: 1;"
          @value-changed="${(event: ComboBoxValueChangedEvent) => {
            this.selectedValue = event.detail.value;
          }}"
        ></vaadin-combo-box>
        <vaadin-button
          theme="primary"
          @click="${() => {
            const person = this.items.find((p) => String(p.id) === this.selectedValue);
            const isInvited = person && this.invitedPeople.some((p) => p.id === person.id);
            if (person && !isInvited) {
              this.invitedPeople = [...this.invitedPeople, person];
              this.selectedValue = '';
            }
          }}"
        >
          Send invite
        </vaadin-button>
      </vaadin-horizontal-layout>

      ${this.invitedPeople.length === 0
        ? this.renderNoInvitationAlert()
        : this.renderInvitedPeopleTable()}
    `;
  }

  private manageRenderer: GridColumnBodyLitRenderer<Person> = ({ id }) => html`
    <vaadin-button
      theme="error tertiary icon"
      @click="${() => {
        this.invitedPeople = this.invitedPeople.filter((p) => p.id !== id);
      }}"
    >
      <vaadin-icon icon="vaadin:trash"></vaadin-icon>
    </vaadin-button>
  `;

  private renderNoInvitationAlert = () => html`
    <div
      style="padding: var(--lumo-size-l);text-align: center;font-style: italic; color: var(--lumo-contrast-70pct);"
    >
      No invitation has been sent
    </div>
  `;

  private renderInvitedPeopleTable = () => html`
    <!-- tag::snippet[] -->
    <vaadin-grid .items="${this.invitedPeople}" all-rows-visible>
      <vaadin-grid-column header="Name" path="displayName" auto-width></vaadin-grid-column>
      <vaadin-grid-column path="email"></vaadin-grid-column>
      <vaadin-grid-column path="address.phone"></vaadin-grid-column>
      <vaadin-grid-column
        header="Manage"
        ${columnBodyRenderer(this.manageRenderer, [])}
      ></vaadin-grid-column>
    </vaadin-grid>
    <!-- end::snippet[] -->
  `;
}

File: src/main/java/com/vaadin/demo/component/grid/GridDynamicHeight.java
package com.vaadin.demo.component.grid;

import java.util.ArrayList;
import java.util.List;

import com.vaadin.demo.domain.Person;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.button.ButtonVariant;
import com.vaadin.flow.component.combobox.ComboBox;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.icon.Icon;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.data.renderer.ComponentRenderer;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;

@Route("grid-dynamic-height")
public class GridDynamicHeight extends Div {

    private static final List<Person> invitedPeople = new ArrayList<>();

    private static Grid<Person> grid;
    private static Div hint;

    public GridDynamicHeight() {
        this.setupInvitationForm();
        this.setupGrid();
        this.refreshGrid();
    }

    private void setupInvitationForm() {
        List<Person> people = DataService.getPeople();
        ComboBox<Person> comboBox = new ComboBox<>();
        comboBox.setItems(people);
        comboBox.setItemLabelGenerator(Person::getFullName);

        Button button = new Button("Send invite");
        button.addThemeVariants(ButtonVariant.LUMO_PRIMARY);
        button.addClickListener(e -> {
            sendInvitation(comboBox.getValue());
            comboBox.setValue(null);
        });

        HorizontalLayout layout = new HorizontalLayout(comboBox, button);
        layout.setFlexGrow(1, comboBox);

        add(layout);
    }

    private void setupGrid() {
        // tag::snippet[]
        grid = new Grid<>(Person.class, false);
        grid.setAllRowsVisible(true);
        // end::snippet[]
        grid.addColumn(Person::getFullName).setHeader("Name");
        grid.addColumn(Person::getEmail).setHeader("Email");
        grid.addColumn(person -> person.getAddress().getPhone())
                .setHeader("Phone");
        grid.addColumn(
                new ComponentRenderer<>(Button::new, (button, person) -> {
                    button.addThemeVariants(ButtonVariant.LUMO_ICON,
                            ButtonVariant.LUMO_ERROR,
                            ButtonVariant.LUMO_TERTIARY);
                    button.addClickListener(e -> this.removeInvitation(person));
                    button.setIcon(new Icon(VaadinIcon.TRASH));
                })).setHeader("Manage");

        grid.setItems(invitedPeople);

        hint = new Div();
        hint.setText("No invitation has been sent");
        hint.getStyle().set("padding", "var(--lumo-size-l)")
                .set("text-align", "center").set("font-style", "italic")
                .set("color", "var(--lumo-contrast-70pct)");

        add(hint, grid);
    }

    private void refreshGrid() {
        if (invitedPeople.size() > 0) {
            grid.setVisible(true);
            hint.setVisible(false);
            grid.getDataProvider().refreshAll();
        } else {
            grid.setVisible(false);
            hint.setVisible(true);
        }
    }

    private void sendInvitation(Person person) {
        if (person == null || invitedPeople.contains(person))
            return;
        invitedPeople.add(person);
        this.refreshGrid();
    }

    private void removeInvitation(Person person) {
        if (person == null)
            return;
        invitedPeople.remove(person);
        this.refreshGrid();
    }

    public static class Exporter // hidden-source-line
            extends DemoExporter<GridDynamicHeight> { // hidden-source-line
    } // hidden-source-line
}
